lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    def determine_agent_level_monitored_resource_via_legacy
+      resource = Google::Apis::LoggingV2beta1::MonitoredResource.new(
+        labels: {})
+      resource.type = determine_agent_level_monitored_resource_type
+      resource.labels = determine_agent_level_monitored_resource_labels(
+        resource.type)
+      resource
+    end
+
+    # Determine agent level monitored resource type.
+    def determine_agent_level_monitored_resource_type
+      # EC2 instance.
+      return EC2_CONSTANTS[:resource_type] if
         @platform == Platform::EC2
-      # See http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
-      open('http://' + METADATA_SERVICE_ADDR +
-           '/latest/dynamic/instance-identity/document') do |f|
-        contents = f.read
-        return JSON.parse(contents)
+
+      # Unknown platform will be defaulted to GCE instance..
+      return COMPUTE_CONSTANTS[:resource_type] if
+        @platform == Platform::OTHER
+
+      # Resource types determined by @subservice_name config.
+      # Cloud Dataflow.
+      return DATAFLOW_CONSTANTS[:resource_type] if
+        @subservice_name == DATAFLOW_CONSTANTS[:service]
+      # Cloud ML.
+      return ML_CONSTANTS[:resource_type] if
+        @subservice_name == ML_CONSTANTS[:service]
+      # Default back to GCE if invalid value is detected.
+      return COMPUTE_CONSTANTS[:resource_type] if
+        @subservice_name
+
+      # Resource types determined by @detect_subservice config.
+      if @detect_subservice
+        begin
+          attributes = fetch_gce_metadata('instance/attributes/').split
+        rescue StandardError => e
+          @log.error 'Failed to detect subservice: ', error: e
+        end
+        # GAE app.
+        return APPENGINE_CONSTANTS[:resource_type] if
+          attributes.include?('gae_backend_name') &&
+          attributes.include?('gae_backend_version')
+        # GKE container.
+        return CONTAINER_CONSTANTS[:resource_type] if
+          attributes.include?('kube-env')
+        # Cloud Dataproc.
+        return DATAPROC_CONSTANTS[:resource_type] if
+          attributes.include?('dataproc-cluster-uuid') &&
+          attributes.include?('dataproc-cluster-name')
+      end
+      # GCE instance.
+      COMPUTE_CONSTANTS[:resource_type]
+    end
+
+    # Determine agent level monitored resource labels based on the resource
+    # type. Each resource type has its own labels that need to be filled in.
+    def determine_agent_level_monitored_resource_labels(type)
+      labels = {}
+
+      case type
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        begin
+          labels['module_id'] = fetch_gce_metadata(
+            'instance/attributes/gae_backend_name')
+          labels['version_id'] = fetch_gce_metadata(
+            'instance/attributes/gae_backend_version')
+        rescue StandardError => e
+          @log.error 'Failed to set monitored resource labels for GAE: ',
```

Comment:
On a second thought, we can just log with the resource type name. So we can achieve having one `begin` and `rescue` without introducing an extra map. PTAL.